test.sh

```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+
+#git fetch origin
+
+function diff_files () {
+    git diff origin/master --name-only
+}
+
+function build_commited () {
+    # Build only commited changed
+    #TODO: mode=sync/async
+    git diff origin/master --name-only | xargs python3 stats.py --build --files
+}
+
+function build_all () {
+    opt="$1"
+    mode="$2" # sync/async
+    if [ "$opt" = "--master-only" ] || [ "$opt" = "-m" ]; then
+        git checkout master
```

Comment:
why we have `--master-only` and `--master-with-commited`?

I think don't makes sense doing `git checkout` here... this can be tricky. You mean `--master-only` to only build the master? You really can do that easily: checkout in the desired branch and using that script... I think we should insert a sensitive git operation inside of this script.

I know that probably things being messy in our discussion, but we should review that (only this part now `./test`). We have another problem here too because the order of arguments matters (the ideal is that doesn't matters, to make things more simple)

Let me be clear, I suggest this CLI design:

* `./test`: build sync and use the diff from master
* `./test sync`: build sync and use the diff from master
* `./test async`: build async and use the diff from master
* `./test [sync|async] diff`: use the diff from master
* `./test [sync|async] staged`: use the diff of staged files (not commited yet) from `git diff --cached --name-only`

And we need a little help function to be printed for `-h` or `--help` flags.

Reply:
If you think we have a better option, we need discuss it. I don't like very much the usage of the current implementation. We need something intuitive what the command do only seeing the command line:

How I can differ `--master-only`, `--commited-only` and `--master-with-commited` without looking to the source? I think this is not clear enough.

Reply:
> why we have --master-only and --master-with-commited?
You mean --master-only to only build the master?

Yes I wasn't focused when writing that ... useless feature

Things are much more clearer now ! ill make the changes now.
One more questions to complete the schema in my head for `./test [sync|async] staged`:
 - can  `git diff --cached --name-only` show different results than `git diff --name-only`?  In PR case ? 
- does `git diff --cached --name-only`  works only if we use `git add [files] ` ?
- are commits in PR considered as stage files for the main repo ?

PS : I already tried to see `git diff --cached --name-only` , output is empty

Reply:
Mmmm maybe it is just for users before even committing in their own fork ?

Reply:
> can git diff --cached --name-only show different results than git diff --name-only? In PR case ?

This is designed always to be executed only on local changes. We have not any result on PR travis system because we don't have `staged` files. If you are not very familiar with git, the `staged` files are after `git add` and before `git commit`. We have only the files after commit them on PR build.

> does git diff --cached --name-only works only if we use git add [files] ?

Yes.

> are commits in PR considered as stage files for the main repo ?

No. Commits on PR is unrelated to stage files, as I said on the first answer, stage files only exists after `git add` and before `git commit`. It's something like a `pre-commit` stage.

> PS : I already tried to see git diff --cached --name-only , output is empty

This is expected if you already commited the changes. The design to use `staged` test is to test your solution before commit them, that way avoiding pain and suffering to create more useless commits to fix sometimes something trivial.

Reply:
> Mmmm maybe it is good for users before even committing in their own fork ?

I don't have sure if I understood that correctly, but I think that users SHOULD test your commits or stage files in own fork.

Reply:
Thank you for answers ! I have better understanding of git diff manipulations.

> I don't have sure if I understood that correctly, but I think that users SHOULD test your commits or stage files in your own fork.

Yes exactly ! Users can be test answers in their own machine before even committing in their fork.